pkg/controller/persistentvolume/persistentvolume_claim_binder_controller.go

```diff
@@ -158,44 +159,56 @@ func syncVolume(volumeIndex *persistentVolumeOrderedIndex, binderClient binderCl
 	currentPhase := volume.Status.Phase
 	nextPhase := currentPhase
 
+	_, exists, err := volumeIndex.Get(volume)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		volumeIndex.Add(volume)
+	}
+
 	switch currentPhase {
-	// pending volumes are available only after indexing in order to be matched to claims.
 	case api.VolumePending:
 		if volume.Spec.ClaimRef != nil {
-			// Pending volumes that have a ClaimRef were recently recycled.  The Recycler set the phase to VolumePending
-			// to start the volume again at the beginning of this lifecycle.
-			// ClaimRef is the last bind between persistent volume and claim.
-			// The claim has already been deleted by the user at this point
-			oldClaimRef := volume.Spec.ClaimRef
-			volume.Spec.ClaimRef = nil
-			_, err = binderClient.UpdatePersistentVolume(volume)
+			claim, err := binderClient.GetPersistentVolumeClaim(volume.Spec.ClaimRef.Namespace, volume.Spec.ClaimRef.Name)
 			if err != nil {
-				// rollback on error, keep the ClaimRef until we can successfully update the volume
-				volume.Spec.ClaimRef = oldClaimRef
-				return fmt.Errorf("Unexpected error saving PersistentVolume: %+v", err)
+				if errors.IsNotFound(err) {
+					// Pending volumes that have a ClaimRef where the claim is missing were recently recycled.
+					// The Recycler set the phase to VolumePending to start the volume at the beginning of this lifecycle.
+					// removing ClaimRef unbinds the volume
+					clone, err := conversion.NewCloner().DeepCopy(volume)
+					if err != nil {
+						return fmt.Errorf("Error cloning pv: %v", err)
+					}
+					volumeClone, ok := clone.(*api.PersistentVolume)
+					if !ok {
+						return fmt.Errorf("Unexpected pv cast error : %v\n", volumeClone)
+					}
+					volumeClone.Spec.ClaimRef = nil
+
+					if updatedVolume, err := binderClient.UpdatePersistentVolume(volumeClone); err != nil {
+						return fmt.Errorf("Unexpected error saving PersistentVolume: %+v", err)
+					} else {
```

Comment:
> it would already descope at the end of the if block.

exactly.  It is therefore unavailable for volumeIndex.Update(updatedVolume)
unless within the else clause.

On Wed, Nov 4, 2015 at 11:09 AM, David Eads <notifications@github.com>
wrote:

> In
> pkg/controller/persistentvolume/persistentvolume_claim_binder_controller.go
> <https://github.com/kubernetes/kubernetes/pull/16432#discussion_r43899635>
> :
>
> > +					// Pending volumes that have a ClaimRef where the claim is missing were recently recycled.
> > +					// The Recycler set the phase to VolumePending to start the volume at the beginning of this lifecycle.
> > +					// removing ClaimRef unbinds the volume
> > +					clone, err := conversion.NewCloner().DeepCopy(volume)
> > +					if err != nil {
> > +						return fmt.Errorf("Error cloning pv: %v", err)
> > +					}
> > +					volumeClone, ok := clone.(*api.PersistentVolume)
> > +					if !ok {
> > +						return fmt.Errorf("Unexpected pv cast error : %v\n", volumeClone)
> > +					}
> > +					volumeClone.Spec.ClaimRef = nil
> > +
> > +					if updatedVolume, err := binderClient.UpdatePersistentVolume(volumeClone); err != nil {
> > +						return fmt.Errorf("Unexpected error saving PersistentVolume: %+v", err)
> > +					} else {
>
> No need for the else.
>
> it would already descope at the end of the if block.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/16432/files#r43899635>.
>